### PR TITLE
dom: Vehicle entity initial commit

### PR DESCRIPTION
**Pull request by vlkolmakov**
This PR amends Worker entity with Vehicle information. Together they represent a logical unit ready for database storage.

@jasonprado please review the Vehicle struct and constants.

@jasonprado, @kevinmichaelchen: **arch questions:**

1. Who is responsible for generating UUID for all entities in the system. API side or OME?
2. What UUID format, if any, is used throughout the system? Is it `[16]byte` type used in PostgreSQL uuid type? If so, we can greatly optimize memory/storage usage for all UUID types in the model.

From what I saw in the bronx repo, the UUID type is not unified and can be in any form, including base64 encoded strings. Am I correct?

### CODE REVIEW

**Comment by kepkap** [comment]
looks good to me. 

points for consideration:

"dom" package name stands for domain object model? 

we have directory core/model containing entities in dedicated packages, i.e. itinerary, job, worker, etc. We need to agree on what approach  to use.  Not a blocker for now, can be refactored anytime, but we bette had it settled earlier.

**Review by kepkap — APPROVED** [review]

**Comment by vlkolmakov** [comment]
> "dom" package name stands for domain object model?

@kepkap, very good question indeed. Yes, `dom` is the domain object model and this package will be eventual flat container for all agreed entities.

Current `core\model` with it's sub-packages per entity role is unnecessary complex, long and WRONG. `core` is not needed at all. `dom` is better than `model` because it shorter and follows general principal of three letter package naming.

Why sub-packaging per model role is WRONG. Because eventually you'll bump into cross referencing among entities in different packages which is prohibited in Go. Like saving Job will require updating Step or vice-versa. So to overcome this you will need to introduce another package which will bring even more complexity to basically simple problem.

We keep the division as it is for now. And see how to handle this issue after all entities are agreed upon.

**Comment by kepkap** [comment]
pros/cons

core/domain/(subpackages):
+ isolation, encapsulation, better handled API public/private design
+ entity model is easier to understand due to limited package content

- i do not see any, except we should not make folder hierarchy too deep. we can get rid of "model" folder and place all entity packages under "core"/(subpackages)   

dom:
+ short named, better for simple models
- provokes repeating prefixes to various model elements: WorkerCreated WorkerApproved, etc. under "worker" package Worker prefix would not be needed
- no isolation in visibility inside dom package, no way to encapsulate private api.


>>>Like saving Job will require updating Step or vice-versa.
is a cyclic dependency, a fundamentally bad design, pure modelling issue. Placing dependant entities into separate package is exact solution to avoid that.

Side note for striving for too short naming. There is no need to use 3-letters acronyms, the code quickly becomes cryptic, and harder to maintain.

Also we should not repeat i.e. itinerary.NewItinerary(), should use GO convention  itinerary.New() . these little flaws are easy to spot and refactor.

**Comment by vlkolmakov** [comment]
> is a cyclic dependency, a fundamentally bad design, pure modelling issue. Placing dependant entities into separate package is exact solution to avoid that.

This is all a book theory. In practice this is not going to work. And how placing dependent entities in separate packages is going to fix this??? They already dependent.

Again, we keep two approaches separate for now. This PR is about Vehicle model.

**Comment by kepkap** [comment]
there should not be mutually dependant entities. placing entities in separate packages forces unidirectional dependency.

**Comment by vlkolmakov** [comment]
> there should not be mutually dependant entities. placing entities in separate packages forces unidirectional dependency.

We'll see when we will have full model in place together with database storage and RPC endpoints.

**Comment by vlkolmakov** [comment]
Right now I am more concerned with the universal UUID format. It will allow us to save a lot of memory and storage.

File: dom/vehicle.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	CarTypeSedan = iota
+	CarTypeHatch
+	CarTypeTwoDoor
+	CarTypeCabrio
+	CarTypeSUV
+	CarTypeVan
+	CarTypeBus
+	CarTypeTrack
+	CarTypeCargoVan
+	CarTypeLemo
```
Comment (kevinmichaelchen) [inline dom/vehicle.go:19]:
```suggestion
	CarTypeLimo
```

**Review by kevinmichaelchen — APPROVED** [review]

File: dom/vehicle.go (hunk 1 above)
Comment (vlkolmakov) [reply dom/vehicle.go]: Done. Thanks.

File: dom/vehicle.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	CarClassExecutive
+	CarClassLuxury
+)
+
+// Vehicle represents car properties
+type Vehicle struct {
+	ID         UUID
+	Driver     UUID // current driver's id, nullable
+	Model      string
+	Maker      string
```
Comment (jasonprado) [inline dom/vehicle.go:37]: Make, not Maker

File: dom/vehicle.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	Color      string
+	Plate      string // license plate
+	Class      CarClass
+	Type       CarType
+	Year       uint16
+	Capacity   uint8 // passenger capacity
+	CargoVol   uint8 // cargo volume
+	WheelChair uint8 // wheelchair accommodation (0: none, 1: one, 2: two, ...)
+	ChildSeats uint8 // number of child seats
```
Comment (jasonprado) [inline dom/vehicle.go:46]: These are good

File: dom/vehicle.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	CarTypeHatch
+	CarTypeTwoDoor
+	CarTypeCabrio
+	CarTypeSUV
+	CarTypeVan
+	CarTypeBus
+	CarTypeTruck
+	CarTypeCargoVan
+	CarTypeLimo
+)
```
Comment (jasonprado) [inline dom/vehicle.go:20]: This is okay for a placeholder. We may need to make this configurable by the client; the definition of a "Black Car" or a "SUV" can vary from market to market.

**Comment by jasonprado** [comment]
re: UUID generation, for Vehicle and Worker and Customer IDs, these are going to be opaque strings provided by the client platform. Things like Steps may need UUIDs to be generated inside of OME.

For Job, I think the ID will also be provided by the client platform. I'm not 100% sure on this yet.

Where we do use UUIDs, I like the xid library/pattern that Kevin mentioned previously.

**Review by jasonprado — APPROVED** [review]

**Comment by kevinmichaelchen** [comment]
https://github.com/rs/xid

**Pull request merged**